mautrix/client/api/modules/media_repository.py

```diff
@@ -121,19 +121,32 @@ async def upload_media(
         if filename:
             query["filename"] = filename
 
+        upload_url = None
+
         if async_upload:
             if mxc:
                 raise ValueError("async_upload and mxc can't be provided simultaneously")
-            mxc = await self.unstable_create_mxc()
+            create_response = await self.unstable_create_mxc()
+            mxc = create_response.content_uri
+            upload_url = create_response.upload_url
 
         path = MediaPath.v3.upload
         method = Method.POST
         if mxc:
             server_name, media_id = self.api.parse_mxc_uri(mxc)
-            path = MediaPath.unstable["fi.mau.msc2246"].upload[server_name][media_id]
-            method = Method.PUT
+            if upload_url is None:
+                path = MediaPath.unstable["fi.mau.msc2246"].upload[server_name][media_id]
+                method = Method.PUT
+            else:
+                path = MediaPath.unstable["fi.mau.msc2246"].upload[server_name][media_id].complete
```

Comment:
Sorry I missed this, we should use a different unstable prefix here to not clash with the original async MSC, `com.beeper.msc3870`.